File: Implementierung/Backend/src/main/java/edu/kit/informatik/dto/mapper/interactions/ParticipantMapper.java
package edu.kit.informatik.dto.mapper.interactions;

import edu.kit.informatik.dto.mapper.IModelDtoMapper;
import edu.kit.informatik.dto.userdata.courses.SessionDto;
import edu.kit.informatik.dto.userdata.interactions.ParticipantDto;
import edu.kit.informatik.dto.userdata.interactions.ParticipantTypeDto;
import edu.kit.informatik.model.User;
import edu.kit.informatik.model.userdata.courses.Course;
import edu.kit.informatik.model.userdata.courses.Session;
import edu.kit.informatik.model.userdata.interactions.Interaction;
import edu.kit.informatik.model.userdata.interactions.Participant;
import edu.kit.informatik.model.userdata.interactions.ParticipantType;
import edu.kit.informatik.repositories.CourseRepository;
import edu.kit.informatik.repositories.InteractionRepository;
import edu.kit.informatik.repositories.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

/**
 * {@link IModelDtoMapper} für {@link Session} and {@link SessionDto}
 *
 * @author uekai
 * @author ugqbo
 * @version 1.0
 */
@Service
public class ParticipantMapper implements IModelDtoMapper<Participant, ParticipantDto> {
    private final ParticipantTypeMapper participantTypeMapper;
    private final UserRepository userRepository;
    private final CourseRepository courseRepository;
    private final InteractionRepository interactionRepository;

    /**
     * Konstruktor zum Erstellen eines Objektes der Klasse
     * @param participantTypeMapper {@link ParticipantTypeMapper}
     * @param userRepository {@link UserRepository}
     * @param courseRepository {@link CourseRepository}
     * @param interactionRepository {@link InteractionRepository}
     */
    @Autowired
    public ParticipantMapper(ParticipantTypeMapper participantTypeMapper, UserRepository userRepository,
                             CourseRepository courseRepository, InteractionRepository interactionRepository) {
        this.participantTypeMapper = participantTypeMapper;
        this.userRepository = userRepository;
        this.courseRepository = courseRepository;
        this.interactionRepository = interactionRepository;
    }

    @Override
    public ParticipantDto modelToDto(Participant participant) {
        List<String> courseIds = new ArrayList<>();
        List<String> interactionIds = new ArrayList<>();
        ParticipantTypeDto participantTypeDto = participantTypeMapper.modelToDto(participant.getParticipantType());

        participant.getCourses().forEach(course -> courseIds.add(course.getId()));
        participant.getInteractions().forEach(interaction -> interactionIds.add(interaction.getId()));

        return new ParticipantDto(
                participant.getId(),
                participant.getUser().getId(),
                participant.getFirstName(),
                participant.getLastName(),
                participantTypeDto,
                courseIds,
                interactionIds,
                participant.isVisible(),
                participant.getCreatedAt(),
                participant.getUpdatedAt()
        );
    }

    @Override
    public List<ParticipantDto> modelToDto(List<Participant> participants) {
        List<ParticipantDto> participantDtos = new ArrayList<>();
        participants.forEach(participant -> participantDtos.add(modelToDto(participant)));

        return participantDtos;
    }

    @Override
    public Participant dtoToModel(ParticipantDto participantDto) {
        User user = userRepository.findUserById(participantDto.getUserId()).orElse(null);
        ParticipantType participantType = participantTypeMapper.dtoToModel(participantDto.getParticipantType());
        
        List<Course> courses = new ArrayList<>();
        List<Interaction> interactions = new ArrayList<>();


        if (participantDto.getCourseIds() != null) {
            participantDto.getCourseIds().forEach(courseId ->
                    courseRepository.findCourseById(courseId).ifPresent(courses::add));
        }
        if (participantDto.getInteractionIds() != null) {
            participantDto.getInteractionIds().forEach(interactionId ->
                    interactionRepository.findInteractionById(interactionId).ifPresent(interactions::add));
        }

        Timestamp updatedAt;

        if (participantDto.getUpdatedAt() == null) {
            updatedAt = participantDto.getCreatedAt();
        } else {
            updatedAt = participantDto.getUpdatedAt();
        }

        Participant participant = new Participant(user, participantDto.getFirstName(), participantDto.getLastName(),
                                participantType, participantDto.isVisible(), participantDto.getCreatedAt(), updatedAt);

        participant.setCourses(courses);
        participant.setInteractions(interactions);

        return participant;
    }

    @Override
    public List<Participant> dtoToModel(List<ParticipantDto> participantDtos) {
        List<Participant> participants = new ArrayList<>();
        participantDtos.forEach(participantDto -> participants.add(dtoToModel(participantDto)));

        return participants;
    }
}

File: Implementierung/frontend/paschi/src/dto/mapper/interactions/ParticipantMapper.ts
import {IModelDtoMapper} from "@/dto/mapper/IModelDtoMapper";
import {Participant} from "@/model/userdata/interactions/Participant";
import {ParticipantDto} from "@/dto/userdata/interactions/ParticipantDto";
import {Course} from "@/model/userdata/courses/Course";
import {Interaction} from "@/model/userdata/interactions/Interaction";
import {UserController} from "@/controller/UserController";
import {ParticipantTypeDto} from "@/dto/userdata/interactions/ParticipantTypeDto";
import {useCourseStore} from "@/store/CourseStore";
import {useInteractionStore} from "@/store/InteractionStore";
import {useStudentStore} from "@/store/ParticipantStore";
import {Student} from "@/model/userdata/interactions/Student";
import {CourseService} from "@/service/CourseService";
import {Teacher} from "@/model/userdata/interactions/Teacher";

export class ParticipantMapper implements IModelDtoMapper<Participant, ParticipantDto> {

  private static mapper: ParticipantMapper = new ParticipantMapper();

  private constructor() {
  }

  static getMapper(): ParticipantMapper {
    return ParticipantMapper.mapper;
  }

  modelToDto(participant: Participant): ParticipantDto {
    const courseIds: string[] = [];
    const interactionIds: string[] = [];
    let particpantType = participant.isTeacher() ? ParticipantTypeDto.TEACHER : ParticipantTypeDto.STUDENT;

    participant.courses.forEach((course: Course) => {
      courseIds.push(course.getId)
    });
    participant.interactions.forEach((interaction: Interaction) => {
      interactionIds.push(interaction.getId)
    });

    return new ParticipantDto(
      participant.getId,
      participant.user.getId,
      participant.createdAt,
      participant.updatedAt,
      participant.firstName,
      participant.lastName,
      particpantType,
      courseIds,
      interactionIds,
      participant.visible
    );
  }

  async dtoToModel(participantDto: ParticipantDto): Promise<Participant> {
    const userController = UserController.getUserController();

    if (participantDto.participantType == ParticipantTypeDto.TEACHER) {
      let teacher = useStudentStore().getTeacher();
      if (teacher == undefined) {
        teacher = new Teacher(
          participantDto.id,
          useStudentStore().getNextId(),
          userController.getUser(),
          userController.getUser().firstName,
          userController.getUser().lastName
        );
        teacher.createdAt = participantDto.createdAt;
        teacher.updatedAt = participantDto.updatedAt;
        useStudentStore().setTeacher(teacher);
      } else if (teacher.updatedAt === participantDto.updatedAt) {
        return teacher;
      }

      const interactions: Interaction[] = [];
      for (const id of participantDto.interactionIds) {
        let interaction = useInteractionStore().getInteraction(id);
        if (interaction !== undefined) {
          interactions.push(interaction);
        }
      }
      teacher.interactions = interactions;

      return teacher;
    } else {
      let participant = useStudentStore().getStudent(participantDto.id);
      if (participant == undefined) {
        participant = new Student(
          participantDto.id,
          0,
          userController.getUser(),
          participantDto.firstName,
          participantDto.lastName
        );
        participant.createdAt = participantDto.createdAt;
        participant.updatedAt = participantDto.updatedAt;
        useStudentStore().addStudent(participant);
      } else {
        participant.firstName = participantDto.firstName;
        participant.lastName = participantDto.lastName;
      }

      const courses: Course[] = [];
      const interactions: Interaction[] = [];
      const courseSerivce = CourseService.getService();

      for (const id of participantDto.courseIds) {
        let course = useCourseStore().getCourse(id);
        if (course == undefined) {
          course = await courseSerivce.getById(id);
        }
        if (course !== undefined) {
          courses.push(course);
        }
      }

      for (const id of participantDto.interactionIds) {
        let interaction = useInteractionStore().getInteraction(id);
        if (interaction !== undefined) {
          interactions.push(interaction);
        }
      }

      participant.courses = courses;
      participant.interactions = interactions;
      participant.visible = participantDto.visible;

      return participant;
    }
  }
}
